src/execution_plans/stage.rs

```diff
@@ -268,6 +259,12 @@ impl ExecutionPlan for StageExec {
         partition: usize,
         context: Arc<TaskContext>,
     ) -> Result<datafusion::execution::SendableRecordBatchStream> {
+        if partition > 0 {
+            return exec_err!(
+                "an executable StageExec must only have 1 partition, but it was called with partition index {partition}"
```

Comment:
Nice safety cache. We want to see this error/panic rather than running a wrong plan.

I wonder if it is useful to add comments somewhere, may be right above this `execute` function, describing something like you did:

> The StageExec node calls try_assing_urls lazily upon calling .execute(). This means that .execute() must only be called once, as we cannot afford to perform several random URL assignation while calling multiple partitions, as they will differ, producing an invalid plan